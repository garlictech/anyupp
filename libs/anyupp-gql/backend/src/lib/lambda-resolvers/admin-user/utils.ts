import { CognitoIdentityServiceProvider, DynamoDB } from 'aws-sdk';
import * as CrudApi from '@bgap/crud-gql/api';
import { from } from 'rxjs';
import { map, tap } from 'rxjs/operators';

export interface AdminUserResolverDeps {
  userPoolId: string;
  cognitoidentityserviceprovider: CognitoIdentityServiceProvider;
  userNameGenerator: () => string;
  docClient: DynamoDB.DocumentClient;
  adminUserTableName: string;
}

export const deleteAdminUserFromTable =
  (deps: AdminUserResolverDeps) =>
  (username: string): ReturnType<CrudApi.CrudSdk['DeleteAdminUser']> =>
    from(
      deps.docClient
        .delete({
          TableName: deps.adminUserTableName,
          Key: {
            id: username,
          },
        })
        .promise(),
    ).pipe(map(res => res.Attributes as CrudApi.Maybe<CrudApi.AdminUser>));
