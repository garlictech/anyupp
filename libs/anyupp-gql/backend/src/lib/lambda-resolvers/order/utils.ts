import { CrudSdk } from '@bgap/crud-gql/api';

export interface OrderResolverDeps {
  crudSdk: CrudSdk;
}
