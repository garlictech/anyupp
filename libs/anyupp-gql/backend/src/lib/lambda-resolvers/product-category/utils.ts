import { CrudSdk } from '@bgap/crud-gql/api';

export interface ProductCategoryResolverDeps {
  crudSdk: CrudSdk;
}
