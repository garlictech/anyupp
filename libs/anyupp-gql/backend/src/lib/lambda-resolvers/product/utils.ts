import { CrudSdk } from '@bgap/crud-gql/api';

export interface ProductResolverDeps {
  crudSdk: CrudSdk;
}
