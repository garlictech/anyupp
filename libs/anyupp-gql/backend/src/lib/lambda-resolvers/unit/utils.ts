import { CrudSdk } from '@bgap/crud-gql/api';

export interface UnitsResolverDeps {
  crudSdk: CrudSdk;
}
