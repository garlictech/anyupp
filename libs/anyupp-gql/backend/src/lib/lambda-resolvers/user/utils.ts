import { AnyuppSdk } from '@bgap/anyupp-gql/api';

export interface UserResolverDeps {
  anyuppSdk: AnyuppSdk;
  consumerUserPoolId: string;
}
