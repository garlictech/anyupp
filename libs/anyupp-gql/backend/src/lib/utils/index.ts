export * from './date.utils';
export * from './dynamodb.utils';
export * from './number.utils';
