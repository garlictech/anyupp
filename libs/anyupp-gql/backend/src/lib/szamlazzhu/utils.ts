import * as CrudApi from '@bgap/crud-gql/api';
import * as R from 'ramda';
import * as fp from 'lodash/fp';
import { flow, pipe } from 'fp-ts/lib/function';

export const getTranslation = (localizedItem: CrudApi.LocalizedItem) =>
  localizedItem['hu'] || localizedItem['en'] || 'ismeretlen termék';

export const calculaterServiceFeeItems = (
  serviceFeePolicy: CrudApi.Maybe<CrudApi.ServiceFeePolicy> | undefined,
  items: CrudApi.OrderItem[],
  currency: string,
): [string, CrudApi.Price][] =>
  pipe(
    items,
    R.map(items => items.serviceFee),
    R.reject(R.isNil),
    x => x as CrudApi.Price[],
    R.groupBy((fee: CrudApi.Price) => fee.taxPercentage.toString()),
    fp.mapValues(
      flow(
        R.map(item => item.netPrice),
        R.sum,
      ),
    ),
    R.toPairs,
    R.map(([taxPercentage, sum]: [string, number]) => [
      parseFloat(taxPercentage),
      sum,
    ]),
    R.map(([taxPercentage, netPrice]) => [
      `Szervizdíj - ${serviceFeePolicy?.percentage}%`,
      {
        taxPercentage,
        netPrice,
        currency,
      },
    ]),
  );
