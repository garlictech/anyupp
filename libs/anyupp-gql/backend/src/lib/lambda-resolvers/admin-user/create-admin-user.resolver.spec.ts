import { createAdminUser } from './create-admin-user.resolver';
import { defer, from, of } from 'rxjs';
import { AdminUserResolverDeps } from './utils';
import { catchError } from 'rxjs/operators';

// We use any-s to mock the system partially

const userName = 'GENERATED_USERNAME';

const goodDeps: AdminUserResolverDeps = {
  userPoolId: 'USER_POOL_ID',
  cognitoidentityserviceprovider: {
    listUsers: jest
      .fn()
      .mockReturnValue({ promise: () => Promise.resolve({ Users: [] }) }),

    adminCreateUser: jest
      .fn()
      .mockReturnValue({ promise: () => Promise.resolve(true) }),

    adminDeleteUser: jest
      .fn()
      .mockReturnValue({ promise: () => Promise.resolve(true) }),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any,
  userNameGenerator: jest.fn().mockReturnValue(userName),
  adminUserTableName: 'ADMIN_USER_TABLE',
  docClient: {
    put: jest.fn().mockReturnValue({
      promise: () => Promise.resolve({ Attributes: 'PUT RETURNED' }),
    }),
  } as any,
};

const createGoodResolverCall = (deps: AdminUserResolverDeps) =>
  defer(() =>
    from(
      createAdminUser({
        input: { email: 'a@a.hu', name: 'Test Elek', phone: '+123456789' },
      })(deps),
    ),
  );

test('Handle the good case', done => {
  createGoodResolverCall(goodDeps).subscribe(res => {
    expect(
      (goodDeps.cognitoidentityserviceprovider.adminCreateUser as jest.Mock)
        .mock.calls[0],
    ).toMatchSnapshot('adminCreateUser parameters');

    expect(
      (goodDeps.cognitoidentityserviceprovider.listUsers as jest.Mock).mock
        .calls[0],
    ).toMatchSnapshot('listUsers parameters');

    expect(
      (goodDeps.docClient.put as jest.Mock).mock.calls[0][0],
    ).toMatchSnapshot(
      {
        Item: {
          createdAt: expect.any(String),
          updatedAt: expect.any(String),
        },
      },
      'CreateAdminUser parameters',
    );

    expect(res).toMatchSnapshot(
      {
        createdAt: expect.any(String),
        updatedAt: expect.any(String),
      },
      'RESULT',
    );
    done();
  });
});

const listUserCaseDeps = (result: Promise<unknown>): AdminUserResolverDeps => ({
  ...goodDeps,
  cognitoidentityserviceprovider: {
    ...goodDeps.cognitoidentityserviceprovider,
    listUsers: jest.fn().mockReturnValue({
      promise: () => result,
    }),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any,
});

test('Handle listUser unknown error', done => {
  const deps = listUserCaseDeps(
    Promise.reject(new Error('UNKNOWN LISTUSERS ERROR')),
  );

  createGoodResolverCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});

test('Handle listUser returns bad data', done => {
  const deps = listUserCaseDeps(Promise.resolve(undefined));

  createGoodResolverCall(deps).subscribe({
    next: res => {
      expect(res).toMatchSnapshot({
        createdAt: expect.any(String),
        updatedAt: expect.any(String),
      });
      done();
    },
  });
});

test('Handle multiple users found error', done => {
  const deps = listUserCaseDeps(Promise.resolve({ Users: [{}, {}] }));

  createGoodResolverCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});

test('Handle user already exists error', done => {
  const deps = listUserCaseDeps(Promise.resolve({ Users: [{}] }));

  createGoodResolverCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});

const adminCreateUserCaseDeps = (
  result: Promise<unknown>,
): AdminUserResolverDeps => ({
  ...goodDeps,
  cognitoidentityserviceprovider: {
    ...goodDeps.cognitoidentityserviceprovider,
    adminCreateUser: jest.fn().mockReturnValue({
      promise: () => result,
    }),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any,
});

test('Handle adminCreateUser unknown error', done => {
  const deps = adminCreateUserCaseDeps(
    Promise.reject(new Error('UNKNOWN ADMINCREATEUSER ERROR')),
  );

  createGoodResolverCall(deps).subscribe({
    error: res => {
      expect(res).toMatchSnapshot();
      done();
    },
  });
});

test('Handle adminCreateUser bad data', done => {
  const deps = adminCreateUserCaseDeps(Promise.resolve(undefined));

  createGoodResolverCall(deps).subscribe({
    next: res => {
      expect(res).toMatchSnapshot({
        createdAt: expect.any(String),
        updatedAt: expect.any(String),
      });
      done();
    },
  });
});

const CrudsdkCreateUserCaseDeps = (
  result: Promise<unknown>,
): AdminUserResolverDeps => ({
  ...goodDeps,
  docClient: {
    ...goodDeps.docClient,
    put: jest.fn().mockReturnValue({ promise: () => result }),
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
  } as any,
});

test('Handle crudSdk.CreateAdminUser error', done => {
  const deps = CrudsdkCreateUserCaseDeps(Promise.reject('dynamodb PUT ERROR'));

  createGoodResolverCall(deps).subscribe({
    error: err => {
      expect(err).toMatchSnapshot();
      done();
    },
  });
});
