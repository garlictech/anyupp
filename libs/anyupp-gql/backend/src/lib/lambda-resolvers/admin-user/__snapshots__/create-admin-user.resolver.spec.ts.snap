// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handle adminCreateUser bad data 1`] = `"UnkownCognitoError"`;

exports[`Handle adminCreateUser unknown error 1`] = `[Error: UNKNOWN ADMINCREATEUSER ERROR]`;

exports[`Handle crudSdk.CreateAdminUser error: adminDeleteUser parameters 1`] = `
Array [
  Object {
    "UserPoolId": "USER_POOL_ID",
    "Username": "GENERATED_USERNAME",
  },
]
`;

exports[`Handle crudSdk.CreateAdminUser error: result 1`] = `
Object {
  "code": "UnknownError",
  "message": "\\"crudSdk.CreateAdminUser ERROR\\"",
}
`;

exports[`Handle listUser returns bad data 1`] = `"UnkownCognitoError"`;

exports[`Handle listUser unknown error 1`] = `[Error: UNKNOWN LISTUSERS ERROR]`;

exports[`Handle multiple users found error 1`] = `
Object {
  "code": "DatabaseError",
  "message": "Something bad happened, more users found with the provided data",
}
`;

exports[`Handle the good case 1`] = `"GENERATED_USERNAME"`;

exports[`Handle the good case: CreateAdminUser parameters 1`] = `
Array [
  Object {
    "input": Object {
      "email": "a@a.hu",
      "id": "GENERATED_USERNAME",
      "name": "Test Elek",
      "phone": "+123456789",
    },
  },
]
`;

exports[`Handle the good case: adminCreateUser parameters 1`] = `
Array [
  Object {
    "DesiredDeliveryMediums": Array [
      "EMAIL",
    ],
    "UserAttributes": Array [
      Object {
        "Name": "email",
        "Value": "a@a.hu",
      },
      Object {
        "Name": "phone_number",
        "Value": "+123456789",
      },
      Object {
        "Name": "name",
        "Value": "Test Elek",
      },
    ],
    "UserPoolId": "USER_POOL_ID",
    "Username": "GENERATED_USERNAME",
  },
]
`;

exports[`Handle the good case: listUsers parameters 1`] = `
Array [
  Object {
    "AttributesToGet": Array [
      "email",
      "phone_number",
    ],
    "Filter": "email = \\"a@a.hu\\"",
    "Limit": 2,
    "UserPoolId": "USER_POOL_ID",
  },
]
`;

exports[`Handle user already exists error 1`] = `
Object {
  "code": "UserAlreadyExists",
  "message": "User already exists",
}
`;
